include/merlin_hashtable.cuh

```diff
@@ -272,84 +309,97 @@ class HashTable {
       if (metas == nullptr) {
         upsert_kernel_with_io<key_type, vector_type, meta_type, DIM, TILE_SIZE>
             <<<grid_size, block_size, 0, stream>>>(
-                table_, keys, reinterpret_cast<const vector_type*>(values),
-                table_->buckets, table_->buckets_size, table_->bucket_max_size,
+                table_, keys, value_vectors, table_->buckets,
+                table_->buckets_size, table_->bucket_max_size,
                 table_->buckets_num, N);
       } else {
         upsert_kernel_with_io<key_type, vector_type, meta_type, DIM, TILE_SIZE>
             <<<grid_size, block_size, 0, stream>>>(
-                table_, keys, reinterpret_cast<const vector_type*>(values),
-                metas, table_->buckets, table_->buckets_size,
-                table_->bucket_max_size, table_->buckets_num, N);
+                table_, keys, value_vectors, metas, table_->buckets,
+                table_->buckets_size, table_->bucket_max_size,
+                table_->buckets_num, N);
       }
     } else {
-      vector_type** d_dst = nullptr;
-      int* d_src_offset = nullptr;
+      auto device_ws = device_memory_pool_->get_workspace<2>(stream);
+      auto d_dst = device_ws.get<vector_type**>(0);
+      auto d_src_offset = device_ws.get<int*>(1);
 
-      CUDA_CHECK(cudaMallocAsync(&d_dst, n * sizeof(vector_type*), stream));
-      CUDA_CHECK(cudaMemsetAsync(d_dst, 0, n * sizeof(vector_type*), stream));
-      CUDA_CHECK(cudaMallocAsync(&d_src_offset, n * sizeof(int), stream));
-      CUDA_CHECK(cudaMemsetAsync(d_src_offset, 0, n * sizeof(int), stream));
+      HostMemoryPool::StaticWorkspace<3> host_ws;
+      vector_type** h_dst;
+      vector_type* h_values;
+      int* h_src_offset;
 
-      {
-        const size_t block_size = options_.block_size;
-        const size_t N = n * TILE_SIZE;
-        const size_t grid_size = SAFE_GET_GRID_SIZE(N, block_size);
+      size_t max_bs;
+      if (options_.io_by_cpu) {
+        host_ws = std::move(host_memory_pool_->get_workspace<3>(stream));
+        h_dst = host_ws.get<vector_type**>(0);
```

Comment:
I know the `(0)`, `(1)` the logic should be work, but it is a little confused, as I recommended in other comments, can we input the `n` directly to the pool/workspace to help the pool API more clear?

Reply:
The reason why I design it like this is to avoid any dynamic memory allocation. Not sure if there is a better choice.

Reply:
With enabling move semantics on the `StaticWorkspace<3>`, there shouldn't be a single memory allocation *(assuming the pool has enough buffers available)*. I thought of making the `StaticWorkspace<...>` compatible with `std::tie` or `std::bind`. But it is kind of ugly due to the variable size. Need a template specialization for each size. Also not sure if it would produce overhead.

Reply:
> I know the logic of `(0)`, `(1)` etc. should work, but it is a little confusing, as I recommended in another comment: can we input the `n` directly to the pool/workspace to help the pool API more clear?

Another thing. You cannot put it a pool creation, because different operations will request different amounts of buffers.